GeneratorInterface/Core/src/WeightHelper.cc

```diff
@@ -0,0 +1,333 @@
+      std::cout << "Looking for LHAPDF info in ID " << lhaidText << std::endl;
+
+    if (!lhaidText.empty()) {
+      try {
+        return std::stoi(lhaidText);
+      } catch (std::invalid_argument& e) {
+        pdfGroup.setIsWellFormed(false);
+      }
+    } else if (!pdfGroup.lhaIds().empty()) {
+      return pdfGroup.lhaIds().back() + 1;
+    } else {
+      if (debug_)
+        std::cout << "Looking up LHAPDF ID from name" << weight.groupname << std::endl;
+      return LHAPDF::lookupLHAPDFID(weight.groupname);
+    }
+    return -1;
+  }
+
+  void WeightHelper::updatePdfInfo(gen::PdfWeightGroupInfo& pdfGroup, const ParsedWeight& weight) const {
+    int lhaid = lhapdfId(weight, pdfGroup);
+    if (debug_)
+      std::cout << "LHAID identified as " << lhaid << std::endl;
+    if (pdfGroup.parentLhapdfId() < 0) {
+      int parentId = lhaid - LHAPDF::lookupPDF(lhaid).second;
+      pdfGroup.setParentLhapdfInfo(parentId);
+      pdfGroup.setUncertaintyType(gen::kUnknownUnc);
+
+      std::string description = "";
+      if (pdfGroup.uncertaintyType() == gen::kHessianUnc)
+        description += "Hessian ";
+      else if (pdfGroup.uncertaintyType() == gen::kMonteCarloUnc)
+        description += "Monte Carlo ";
+      description += "Uncertainty sets for LHAPDF set " + LHAPDF::lookupPDF(parentId).first;
+      description += " with LHAID = " + std::to_string(parentId);
+      description += "; ";
+
+      pdfGroup.appendDescription(description);
+    }
+    // after setup parent info, add lhaid
+    pdfGroup.addLhaid(lhaid);
+  }
+
+  void WeightHelper::updatePartonShowerInfo(gen::PartonShowerWeightGroupInfo& psGroup,
+                                            const ParsedWeight& weight) const {
+    if (psGroup.nIdsContained() == DEFAULT_PSWEIGHT_LENGTH) {
+      psGroup.setIsWellFormed(true);
+      psGroup.cacheWeightIndicesByLabel();
+    }
+    if (weight.content.find(':') != std::string::npos && weight.content.find('=') != std::string::npos)
+      psGroup.setNameIsPythiaSyntax(true);
+  }
+
+  bool WeightHelper::splitPdfWeight(ParsedWeight& weight, WeightGroupInfoContainer& weightGroups) const {
+    if (weightGroups[weight.wgtGroup_idx]->weightType() == gen::WeightType::kPdfWeights) {
+      auto& pdfGroup = *static_cast<gen::PdfWeightGroupInfo*>(weightGroups[weight.wgtGroup_idx].get());
+      int lhaid = lhapdfId(weight, pdfGroup);
+      if (lhaid > 0 && !pdfGroup.isIdInParentSet(lhaid) && pdfGroup.parentLhapdfId() > 0) {
+        weightGroups.push_back(buildGroup(weight));
+        weight.wgtGroup_idx++;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void WeightHelper::cleanupOrphanCentralWeight(WeightGroupInfoContainer& weightGroups) const {
+    auto centralIt = std::find_if(std::begin(weightGroups), std::end(weightGroups), [](auto& entry) {
+      return entry->weightType() == gen::WeightType::kScaleWeights &&
+             static_cast<ScaleWeightGroupInfo*>(entry.get())->containsCentralWeight();
```

Comment:
I think the logic here is incorrect: this should select for `!containsCentralWeight()`, in order to find the scale weight group that is *missing* the central weight (because it was orphaned). (I had to change this when testing privately, or the resulting scale weight group did not qualify as well formed.)